File: raw_resources/choice_picker.js
function num_digits(x)
{
    let ans=0;
    while(x>=1)
    {
        ans++;
        x=x/10;
    }
    return ans;
}
function to_int(str) //str is a string of numbers
{
    let ans=0;
    let j=0;
    let n=str.length;
    while(j<n)
    {
        ans=(ans*10)+Number(str[j]);
        print(ans);
        j++;
    }
    return ans;
}
function random_number()
{
    let x=Math.random();
    let y=Math.pow(10,15);
    return Math.floor(x*y);
}
function random_number_between(a,b)
{
    if(a>b)
    {
        let temp=a;
        a=b;
        b=temp;
    }
    let size=(b-a+1);
    let val=random_number();
    val=val%size;
    val=val+a;
    return val;
}
function array_sum(arr,n)
{
    let sum=0;
    let j=0;
    while(j<n)
    {
        sum=sum+arr[j];
        j++;
    }
    return sum;
}
function choice_picker(weights,n) //Input is array of integers and length n
{
    let size=array_sum(weights,n);
    let pick=random_number_between(1,size);
    let j=0;
    let nsum=0;
    while(j<n && nsum<pick)
    {
        nsum=nsum+weights[j];
        j++;
    }
    j--;
    return j;
}


function choice_picker_strin(str_weights,n) // Use this function if input is array of strings (of numbers) instead of array of integers
{
    let weights=[];
    let j=0;
    while(j<n)
    {
        weights[j]=Number(str_weights[j]);
        j++;
    }
    return choice_picker(weights,n);
}



let arr=new Array(1000,2,1000);
print(choice_picker(arr,3)+1);

File: raw_resources/choice-picker.cpp
#include <stdio.h>
#include <bits/stdc++.h>
void initialize_random()
{
    srand(time(0));
}
int get_random_range(int a,int b)
{
    if(a>b)
    b=a;
    int x=rand();
    int len=b-a+1;
    int ans=x%len;
    ans=ans+a;
    return ans;

}
long long array_sum(int* arr,int n)
{
    long long sum=0;
    int j=0;
    while(j<n)
    {
        sum=sum+(long long)arr[j];
        j++;
    }
    return sum;
}
void scan_array(int* arr,int n)
{
    int j=0;
    while(j<n)
    {
        scanf("%d ",arr+j);
        j++;
    }
}
int choice_picker(int* weights,int n)
{
    long long sum=array_sum(weights,n);
    int pick=get_random_range(1,sum);
    printf("pick : %d\n",pick);
    int j=0;
    long long nsum=0;
    while(j<n && nsum<pick)
    {
        nsum=nsum+weights[j];
        j++;
    }
    j--;
    return j;
    
}
int main()
{
    initialize_random();
    int n;
    scanf("%d ",&n);
    int arr[n];
    scan_array(arr,n);
    int x=choice_picker(arr,n);
    printf("\nchoice is : %d\n",x+1);
}